src/store/login/login.thunks.ts

```diff
@@ -0,0 +1,306 @@
+      ) {
+        hermez.Providers.setProvider(process.env.REACT_APP_HARDWARE_WALLETS_PROVIDER);
+      }
+
+      if (walletName === loginActions.WalletName.WALLET_CONNECT) {
+        const walletConnectProvider = new WalletConnectProvider({
+          infuraId: process.env.REACT_APP_INFURA_API_KEY,
+          bridge: process.env.REACT_APP_WALLETCONNECT_BRIDGE,
+        });
+        hermez.Providers.setProvider(walletConnectProvider, hermez.Providers.PROVIDER_TYPES.WEB3);
+      }
+
+      const provider = hermez.Providers.getProvider();
+
+      dispatch(loginActions.loadWallet());
+
+      if (walletName === loginActions.WalletName.METAMASK) {
+        try {
+          await provider.send("eth_requestAccounts", []);
+        } catch (err) {
+          console.error(err);
+        }
+      }
+
+      const signerData = await getSignerData(provider, walletName, accountData);
+      const signer = await hermez.Signers.getSigner(provider, signerData);
+
+      if (provider.provider instanceof WalletConnectProvider) {
+        // Enable shows the QR or uses the stored session
+        await provider.provider.enable();
+      }
+
+      const { chainId, name: chainName } = await provider.getNetwork();
+
+      if (process.env.REACT_APP_ENV === "production" && !isEnvironmentSupported(chainId)) {
+        dispatch(
+          globalActions.openSnackbar("Please, switch your network to Mainnet or Rinkeby to login")
+        );
+        dispatch(loginActions.goToWalletSelectorStep());
+
+        if (provider.provider instanceof WalletConnectProvider) {
+          // Close the stored session to avoid storing a network not supported by Hermez
+          await provider.provider.disconnect();
+        }
+
+        return;
+      }
+
+      dispatch(globalThunks.setHermezEnvironment(chainId, chainName));
+
+      const address = await signer.getAddress();
+      const hermezAddress = hermez.Addresses.getHermezAddress(address);
+      const providerOrSigner =
+        walletName === loginActions.WalletName.WALLET_CONNECT ? provider : signer;
+      const signature = await signMessageHelper(
+        providerOrSigner,
+        hermez.Constants.METAMASK_MESSAGE,
+        address
+      );
+      const hashedSignature = keccak256(signature);
+      const signatureBuffer = hermez.Utils.hexToBuffer(hashedSignature);
+      const wallet = new hermez.HermezWallet.HermezWallet(signatureBuffer, hermezAddress);
+      const {
+        login: { step },
+      } = getState();
+
+      if (step.type === "wallet-loader") {
+        dispatch(globalActions.loadWallet(wallet));
+        // ToDo: SignerData does not have the prop "address" on all its shapes. Why do we need to add the address?
+        //       For now I'm creating an IIFE to set the address where it makes sense, but we should confirm whether
+        //       this is necessary or not. Am I messing it up with the wrong types?
+        const signerDataWithAddress: SignerData = ((): SignerData => {
```

Comment:
> Mmmm, I think that the `address` is not needed in the store, we only need to hold the `SignerData`

Agreed. We only save `SignerData` on the global store.

> Here you can find the data needed in `hermezjs` for each type, if you take a look only the `TREZOR` case needs the address.

Yep, from there I got the info to define the type [`SignerData`](https://github.com/hermeznetwork/wallet-ui/blob/refactor-to-typescript-slice-04/src/hermezjs-typings.d.ts#L665). Notice that the member `JsonRpcSignerData` of the union can also hold the address on its `addressOrIndex` prop. Both props are optional though.

> so maybe it's better to build properly the `signerData` in the `getSignerData` function

I'm not sure I'm getting what you propose here. To properly build the `SignerData` (i.e include the address) in the `getSignerData` function, we need to pass the address to the function. To pass the address to the function we would need to call `getSignerData` twice, one initial time without address and then again with the address. Is that what you mean?

I may be lost, but otherwise, I think that the current implementation (IIFE) is better.

> to later store the address in Redux

Lost here.

Maybe we can meet about this...